pkg/services/ssosettings/api/api.go

```diff
@@ -86,12 +90,21 @@ func (api *Api) getProviderSettings(c *contextmodel.ReqContext) response.Respons
 		return response.Error(http.StatusNotFound, "The provider was not found", err)
 	}
 
-	dto, err := settings.ToSSOSettingsDTO()
+	etag := generateSHA1ETag(settings.OAuthSettings)
+
+	dto, err := settings.ToSSOSettingsDTO(c.QueryBool("includeDefaults"))
 	if err != nil {
 		return response.Error(http.StatusInternalServerError, "The provider is invalid", err)
 	}
 
-	return response.JSON(http.StatusOK, dto)
+	return response.JSON(http.StatusOK, dto).SetHeader("ETag", etag)
+}
+
+func generateSHA1ETag(OAuthSettings *social.OAuthInfo) string {
```

Comment:
Probably it's better to use an SSOSettingsDTO as a parameter and use the `Settings` field of it to support SAML settings in the future.